File: src/main/java/io/pageobject/generator/locator/AbstractLocatorPartGenerator.java
package io.pageobject.generator.locator;

import io.pageobject.generator.GeneratorContext;
import org.jsoup.nodes.Element;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Strings.isNullOrEmpty;
import static io.pageobject.generator.Expressions.replaceIndexBindingWithFunctionParameter;
import static io.pageobject.generator.locator.LocatorPart.*;

public abstract class AbstractLocatorPartGenerator implements LocatorPartGenerator {

    public static final String ELEMENT_FINDER = "element";
    public static final String ELEMENT_ARRAY_FINDER = "element.all";
    public static final String ELEMENT_ARRAY_FINDER_IN_REPEATER = "all";

    private final LocatorSource[] locatorCandidates;
    protected final Element element;

    public AbstractLocatorPartGenerator(Element element, LocatorSource... locatorCandidates) {
        checkArgument(locatorCandidates.length > 0, "There must be at least one locator candidate");
        this.element = element;
        this.locatorCandidates = locatorCandidates;
    }

    @Override
    public LocatorPart firstPart(GeneratorContext context) {
        return invalidPart();
    }

    @Override
    public LocatorPart middlePart(GeneratorContext context, int index) {
        return invalidPart();
    }

    @Override
    public LocatorPart lastPartForRepeaterElement(GeneratorContext context, int index) {
        return emptyPart();
    }

    protected LocatorPart generateLastPart(GeneratorContext context, String elementFinder, int depth) {
        Element element = context.getElement();

        for (LocatorSource locatorCandidate : locatorCandidates) {
            String attributeValue = locatorCandidate.extractLocatorValue(element, context);

            if (!isNullOrEmpty(attributeValue) && context.isLocatorAvailable(locatorCandidate, attributeValue)) {
                StringBuilder value = new StringBuilder(elementFinder);
                value.append("(");
                value.append((locatorCandidate.locator(replaceIndexBindingWithFunctionParameter(attributeValue,
                                                                                                depth))));
                value.append(")");

                return part(value.toString(), locatorCandidate);
            }
        }

        return invalidPart();
    }

    @Override
    public Element getElement() {
        return element;
    }
}

File: lib/locator/abstract-locator-part-generator.js
"use strict";

const assert = require('assert');
const _ = require('lodash');
const invalidPart = require('./locator-part').invalidPart;
const emptyPart = require('./locator-part').emptyPart;
const part = require('./locator-part').part;
const replaceIndexBindingWithFunctionParameter = require('../expressions').replaceIndexBindingWithFunctionParameter;

const elementFinder = "element";
const elementArrayFinder = "element.all";
const elementArrayFinderInRepeater = "all";

class AbstractLocatorPartGenerator {

  constructor(element, ...locatorCandidates) {
    if (this.onlyPart === undefined) {
      throw new TypeError('Must override method onlyPart');
    }

    if (this.lastPart === undefined) {
      throw new TypeError('Must override method lastPart');
    }

    assert(locatorCandidates.length > 0, 'There must be at least one locator candidate');
    this._element = element;
    this._locatorCandidates = locatorCandidates;
  }

  firstPart(context) {
    return invalidPart();
  }

  middlePart(context, index) {
    return invalidPart();
  }

  lastPartForRepeaterElement(context, index) {
    return emptyPart();
  }

  generateLastPart(context, elementFinder, depth) {
    let element = context.element;

    for (let locatorCandidate of this._locatorCandidates) {
      let attributeValue = locatorCandidate.extractLocatorValue(element, context);

      if (!_.isEmpty(attributeValue) && context.isLocatorAvailable(locatorCandidate, attributeValue)) {
        let replacedValue = replaceIndexBindingWithFunctionParameter(attributeValue, depth);
        let locator = locatorCandidate.locator(replacedValue);

        return part(`${elementFinder}(${locator})`, locatorCandidate);
      }
    }

    return invalidPart();
  }

  get element() {
    return this._element;
  }

  static get elementFinder() {
    return elementFinder;
  }

  static get elementArrayFinder() {
    return elementArrayFinder;
  }

  static get elementArrayFinderInRepeater() {
    return elementArrayFinderInRepeater;
  }

}

module.exports = AbstractLocatorPartGenerator;